### PR TITLE
Initial implementation of training/testing loop

### LINKED ISSUE #3: Training/testing functions
We should have easy to use training/testing functions for our model where we pass in e.g. the model, the optimizer, number of epochs, etc.

### PR DESCRIPTION
**Pull request by charlesoblack**
Closes #3. Depends on #9, #23 since it uses an API that I'm not sure is going to look exactly like this.

This introduces a dataset class that is not used - it's an initial implementation (that frankly, I'm not sure it works perfectly) for datasets that are written to files.

The training loop is essentially:

- generate data
- get subset of the datapoints
- train model on that subset
- eval if model is better than old model by self-play

Since I'm not confident on how the MCTS/model APIs are going to work exactly, this'll have to be revisited. I also haven't added any tests yet for the same reason.

One thing I have to look into is whether the loss calling method is usually (output, target) or (target, output). In sklearn, it's usually the latter; in pytorch it seems to be the former.

### CODE REVIEW

**Comment by charlesoblack** [comment]
@abhon can you elaborate on the neural net API here? what does it take as input and what does it output?

**Comment by abhon** [comment]
The main network takes a tenor of the following dimensions:

number of samples * board state(previous moves + player positions) * game length * game height.

In the Alphago Zero paper its game length * game height *  board state, resulting in 19*19*17, but to work with pytorch's conv2d I had to change it a little bit.

It then splits into the PolicyHead, which returns a vector of length game length * game height of move probabilities, and then to the ValueHead, which just returns a number(float between -1 and 1).

**Comment by PhilipEkfeldt** [comment]
> The main network takes a tenor of the following dimensions:
> 
> number of samples * board state(previous moves + player positions) * game length * game height.
> 
> In the Alphago Zero paper its game length * game height * board state, resulting in 19_19_17, but to work with pytorch's conv2d I had to change it a little bit.
> 
> It then splits into the PolicyHead, which returns a vector of length game length * game height of move probabilities, and then to the ValueHead, which just returns a number(float between -1 and 1).

I need to go back to the paper, but is the policyhead always a 1D vector or is it sometimes a 2/3D tensor (excluding batch dimensions) to account for spatial dimensions and different pieces? Just thinking from an API standpoint where we want to transform the policy to better manage the state change based on the action.

File: alphazero/_utils.py [hunk 1]
```diff
@@ -26,3 +31,117 @@ def AlphaGoZeroLoss(
...
+    device=torch.device("cpu"),
+    n_test_games=9,
+):
+
+    data = []
+    # generate data through self-play with MCTS
+    for i in range(n_games):
+        data.append(
+            self_play(
+                game, model, n_mcts_iter=n_mcts_iter, temperature=temperature
```
Comment (PhilipEkfeldt) [inline alphazero/_utils.py:82]: One game will return multiple datapoints (one for every step, returns a list of datapoints in the form of tuples)

File: alphazero/_utils.py [hunk 2]
```diff
@@ -26,3 +31,117 @@ def AlphaGoZeroLoss(
...
+def test(game_type, model_1, model_2, n_games=9, device=torch.device("cpu")):
+    tally = 0
+
+    for game in range(n_games):
+        tally += play_game(game_type, model_1, model_2, device)
+
+    if tally > 0:
+        return model_1
+    # implicit else: if the model_1 ties with the model_2, we keep model_2
+    return model_2
```
Comment (PhilipEkfeldt) [inline alphazero/_utils.py:127]: Here for logging we should make it clear which model won and what the score was, not just return the model

File: alphazero/_utils.py [hunk 3]
```diff
@@ -26,3 +31,117 @@ def AlphaGoZeroLoss(
...
+
+    game = game_type()
+    model_1.to(device)
+    model_2.to(device)
+    current_player = model_1
+
+    while not game_type.is_game_over(game.board_state):
+        # play the moves individually
+        pos = game.board_state.to(device)
+        moves, q = current_player(pos)
```
Comment (PhilipEkfeldt) [inline alphazero/_utils.py:140]: See my comment in the main thread about what the net returns and how that interacts with the game api

File: alphazero/_utils.py [hunk 4]
```diff
@@ -26,3 +31,117 @@ def AlphaGoZeroLoss(
...
+        game.make_move(np.argmax(moves))
+
+        # i'm not confident this works since it relies on "is"
+        current_player = model_2 if current_player is model_1 else model_1
```
Comment (PhilipEkfeldt) [inline alphazero/_utils.py:144]: Think this should work?

**Comment by abhon** [comment]
> > The main network takes a tenor of the following dimensions:
> > number of samples * board state(previous moves + player positions) * game length * game height.
> > In the Alphago Zero paper its game length * game height * board state, resulting in 19_19_17, but to work with pytorch's conv2d I had to change it a little bit.
> > It then splits into the PolicyHead, which returns a vector of length game length * game height of move probabilities, and then to the ValueHead, which just returns a number(float between -1 and 1).
> 
> I need to go back to the paper, but is the policyhead always a 1D vector or is it sometimes a 2/3D tensor (excluding batch dimensions) to account for spatial dimensions and different pieces? Just thinking from an API standpoint where we want to transform the policy to better manage the state change based on the action.

So the paper mentions it as a 1D vector for the policy, but I think it would be easy to rearrange it as the board square with a reshape? Might be easier to store each policy as a vector though and use some arithmetic to find the probability at a certain position.

File: alphazero/_utils.py (hunk 1 above)
Comment (charlesoblack) [reply alphazero/_utils.py:82]: Gotcha. Will you ping me once you have the API for self_play defined?

File: alphazero/_utils.py (hunk 4 above)
Comment (charlesoblack) [reply alphazero/_utils.py:144]: I'm not confident because "is" works weirdly sometimes. I'd like to run a test before actually merging into master.
